Continued cleanup of old UNALIGNED_OK checks

Dead2 opened the pull request:
- Remove obsolete checks
- Fix checks that are inconsistent
- Stop compiling a bunch of compare256/longest_match variants that never gets called
- Moved the generic compare256 functions into a separate file `compare256_p.h`, to make tests/benchmarks possible in a clean way
- Rename functions to get rid of misleading _unaligned names
- Allow overriding OPTIMAL_CMP define

In present develop, we compile generic functions that we never use, 8 variants of longest_match and 5 of compare256, without counting the sse and such optimized versions.
With this PR we only compile in the 2 longest_match and the 2 variants of compare256 that we actually make pointers to use.

## 64-bit platforms
This PR reduces the size of the compiled library by quite a bit by not compiling code that cannot be accessed.
x86_64:  -9984 bytes, so approx 6.9% smaller.
Aarch64: -8600 bytes, so approx 7.3% smaller.

Performance difference on x86-64/Aarch64 is not measurable, because this does not make changes to the code they execute, it only removes dead code.

## 32-bit platforms
Benchmarks of 32-bit should be different, and we do need benchmarks to test those.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced new memory comparison functions for 8, 16, 32, and 64-bit comparisons.
	- Added benchmarks for aligned comparisons.
	- Added new header files to support enhanced functionality.

- **Bug Fixes**
	- Streamlined handling of memory comparisons and removed unnecessary checks.

- **Documentation**
	- Updated comments and documentation to reflect new function logic and benchmarks.

- **Chores**
	- Enhanced build configuration for improved architecture support and clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Dead2 committed:
Continued cleanup of old UNALIGNED_OK checks
- Remove obsolete checks
- Fix checks that are inconsistent
- Stop compiling compare256/longest_match variants that never gets called
- Improve how the generic compare256 functions are handled.
- Allow overriding OPTIMAL_CMP

This simplifies the code and avoids having a lot of code in the compiled library than can never get executed.

Dead2 commented on arch/generic/generic_functions.h:
```diff
@@ -27,45 +27,35 @@ void     inflate_fast_c(PREFIX3(stream) *strm, uint32_t start);
+uint32_t compare256_unaligned_c(const uint8_t *src0, const uint8_t *src1);
+#endif
+#if BYTE_ORDER == LITTLE_ENDIAN
+#  ifdef HAVE_BUILTIN_CTZLL
+        uint32_t longest_match_unaligned_64(deflate_state *const s, Pos cur_match);
         uint32_t longest_match_slow_unaligned_64(deflate_state *const s, Pos cur_match);
-#    endif
+        uint32_t compare256_unaligned_64(const uint8_t *src0, const uint8_t *src1);
+#  elif defined(HAVE_BUILTIN_CTZ)
+        uint32_t longest_match_unaligned_32(deflate_state *const s, Pos cur_match);
+        uint32_t longest_match_slow_unaligned_32(deflate_state *const s, Pos cur_match);
+        uint32_t compare256_unaligned_32(const uint8_t *src0, const uint8_t *src1);
+#  else
+    uint32_t longest_match_unaligned_16(deflate_state *const s, Pos cur_match);
+    uint32_t longest_match_slow_unaligned_16(deflate_state *const s, Pos cur_match);
+    uint32_t compare256_unaligned_16(const uint8_t *src0, const uint8_t *src1);
 #  endif
+#endif
```
Those functions are templated using LONGEST_MATCH macro for naming and including the file match_tpl.h

Dead2 replied:
I am not sure writing to me in Spanish is going to be very helpful.

Dead2 commented on test/test_compare256.cc:
```diff
@@ -12,6 +12,10 @@
+
+// Unorthodox workaround to get access to all functions
+#  define ZTEST_ALL
+#  include "arch/generic/compare256_c.c"
```
This is needed to be able to unit-test the functions in that file, it is an unorthodox workaround.

Dead2 commented:
Rebased, but there are still a few small things to fix, I have to get some Christmas shopping done first though.

ccawley2011 commented on arch/generic/compare256_c.c:
```diff
@@ -8,59 +8,8 @@
-/* 16-bit unaligned integer comparison */
-static inline uint32_t compare256_unaligned_16_static(const uint8_t *src0, const uint8_t *src1) {
+/* 16-bit integer comparison */
+static inline uint32_t compare256_16_static(const uint8_t *src0, const uint8_t *src1) {
```
It might be better to keep the byte version rather than the 16-bit version - the byte version would potentially have better performance than the 16-bit version on platforms without unaligned reads, and platforms without dedicated clz/ctz instructions would likely end up using the 32-bit or 64-bit versions anyway because GCC typically provides generic fallbacks for `__builtin_clz`/`__builtin_ctz` instructions for all architectures.

nmoinvaz replied:
I think I would have to agree, that keeping the c versions is probably a good idea for those platforms.

Dead2 replied:
Added the 8-bit version back, but not currently in use by any platform.

nmoinvaz replied:
The code you have now is a good compromise.

ccawley2011 reviewed (commented):
How would these changes affect performance on platforms without unaligned memory access?

Dead2 commented:
> How would these changes affect performance on platforms without unaligned memory access?

That is something I want some help testing.
Hopefully it is not very a big loss, but we can add back restrictions to that if it is worth it. Most of the cleanup is not contingent on that, but it did look like something that should be investigated.

In my godbolt tests, it looks like the emulated `__builtin_clz` should be faster than the old _c. It adds only a few more instructions, but loses a lot more in longest_match.
But as of yet, this is just a hypothesis that we hope is reflected in actual benchmarks as well.

Dead2 commented:
Oh, if we find any platforms that actually needs the byte-by-byte functions, they should be specifically added with `#define OPTIMAL_CMP 8`, so that we can optimize for it and also document it for posterity.

KungFuJesus commented:
> Oh, if we find any platforms that actually needs the byte-by-byte functions, they should be specifically added with `#define OPTIMAL_CMP 8`, so that we can optimize for it and also document it for posterity.

I'm somewhat speculating but I wonder if it'll still be faster for the compiler to generate a piecewise load by finding partial alignment and shifting and or'ing it in. It'd be a long instruction sequence to be generic enough for any alignment but it still might be better.

Then again on ppc, at least ppc64, a good portion of that time was external to that comparison loop. I guess SPARC will be the ultimate test.

Dead2 commented:
In present develop, we compile generic functions that we never use, 8 variants of longest_match and 5 of compare256, without counting the sse and such optimized versions.
With this PR we only compile in the 2 longest_match and the 2 variants of compare256 that we actually make pointers to use.

nmoinvaz commented on arch/generic/compare256_p.h:
```diff
@@ -0,0 +1,90 @@
+    uint32_t len = 0;
+
+    do {
+        if (zng_memcmp_2(src0, src1) != 0)
+            return len + (*src0 == *src1);
+        src0 += 2, src1 += 2, len += 2;
+
+        if (zng_memcmp_2(src0, src1) != 0)
+            return len + (*src0 == *src1);
+        src0 += 2, src1 += 2, len += 2;
+
+        if (zng_memcmp_2(src0, src1) != 0)
+            return len + (*src0 == *src1);
+        src0 += 2, src1 += 2, len += 2;
+
+        if (zng_memcmp_2(src0, src1) != 0)
+            return len + (*src0 == *src1);
+        src0 += 2, src1 += 2, len += 2;
```
No, it will hurt performance.

Dead2 committed:
Rename functions to get rid of old and now misleading "unaligned" naming

Dead2 commented:
Rebased, reordered and redid the commits.

Also added back the 8-bit compare256 variants. Those are now available in the benchmarks, and can be accessed by setting the define `OPTIMAL_CMP=8` from cflags.
Platforms don't currently use that by default now. But specific platforms can easily be set to use it after showing a performance benefit.

Dead2 commented:
Updated the PR message with current results and status.

We need to do benchmarks of 32-bit library to verify that there is not a big performance drop.

Dead2 commented:
### Compare256 comparison benchmarks
#### Aarch64
```
compare256/8/1              1.04 ns         1.04 ns    671640955
compare256/8/8              4.70 ns         4.70 ns    149090137
compare256/8/64             35.3 ns         35.3 ns     19811925
compare256/8/256             141 ns          141 ns      4946817
compare256/16/1             1.46 ns         1.46 ns    479844439
compare256/16/8             2.94 ns         2.94 ns    238082822
compare256/16/64            19.2 ns         19.2 ns     36509961
compare256/16/256           74.2 ns         74.2 ns      9434937
compare256/32/1             1.25 ns         1.25 ns    559813810
compare256/32/8             2.08 ns         2.08 ns    335891412
compare256/32/64            15.4 ns         15.4 ns     45390532
compare256/32/256           55.4 ns         55.4 ns     12627208
compare256/64/1             1.39 ns         1.39 ns    503839582
compare256/64/8             1.39 ns         1.39 ns    503836488
compare256/64/64            8.75 ns         8.75 ns     79974307
compare256/64/256           28.8 ns         28.8 ns     24340080
compare256/neon/1           2.50 ns         2.50 ns    279910471
compare256/neon/8           2.50 ns         2.50 ns    279911094
compare256/neon/64          7.50 ns         7.50 ns     93302510
compare256/neon/256         27.5 ns         27.5 ns     25446310
compare256_rle/8/1          1.25 ns         1.25 ns    559822680
compare256_rle/8/8          3.75 ns         3.75 ns    186606382
compare256_rle/8/64         30.0 ns         30.0 ns     23325833
compare256_rle/8/256         120 ns          120 ns      5831478
compare256_rle/16/1         1.67 ns         1.67 ns    437495800
compare256_rle/16/8         2.29 ns         2.29 ns    305031052
compare256_rle/16/64        16.7 ns         16.7 ns     41986756
compare256_rle/16/256       66.7 ns         66.7 ns     10496551
compare256_rle/32/1         1.25 ns         1.25 ns    559814227
compare256_rle/32/8         3.75 ns         3.75 ns    186607110
compare256_rle/32/64        15.4 ns         15.4 ns     45390700
compare256_rle/32/256       60.9 ns         60.9 ns     11503274
compare256_rle/64/1         1.46 ns         1.46 ns    479843831
compare256_rle/64/8         1.46 ns         1.46 ns    479846390
compare256_rle/64/64        7.09 ns         7.09 ns     98792587
compare256_rle/64/256       27.1 ns         27.1 ns     25837734
```

#### x86-64
```
compare256/8/1              1.11 ns         1.11 ns    629446602
compare256/8/8              3.06 ns         3.06 ns    228928759
compare256/8/64             20.6 ns         20.6 ns     34021723
compare256/8/256            80.7 ns         80.6 ns      8682776
compare256/16/1            0.838 ns        0.836 ns    836854358
compare256/16/8             1.39 ns         1.39 ns    503523412
compare256/16/64            9.19 ns         9.17 ns     76287725
compare256/16/256           35.9 ns         35.9 ns     19520088
compare256/32/1            0.836 ns        0.834 ns    839312948
compare256/32/8             1.11 ns         1.11 ns    629461873
compare256/32/64            6.13 ns         6.12 ns    114376186
compare256/32/256           22.8 ns         22.8 ns     30709359
compare256/64/1            0.835 ns        0.834 ns    835365938
compare256/64/8            0.835 ns        0.834 ns    835339052
compare256/64/64            4.18 ns         4.18 ns    167534937
compare256/64/256           11.7 ns         11.7 ns     59878910
compare256/sse2/1           1.67 ns         1.67 ns    419685834
compare256/sse2/8           1.67 ns         1.67 ns    419690605
compare256/sse2/64          3.90 ns         3.89 ns    179835170
compare256/sse2/256         13.4 ns         13.3 ns     52326816
compare256/avx2/1           1.95 ns         1.95 ns    359742226
compare256/avx2/8           1.95 ns         1.95 ns    359614229
compare256/avx2/64          2.23 ns         2.22 ns    314774354
compare256/avx2/256         4.28 ns         4.28 ns    163167815
compare256_rle/8/1         0.731 ns        0.730 ns    842684888
compare256_rle/8/8          2.78 ns         2.78 ns    251584239
compare256_rle/8/64         20.3 ns         20.3 ns     34495143
compare256_rle/8/256        80.5 ns         80.3 ns      8712754
compare256_rle/16/1        0.896 ns        0.895 ns    927267201
compare256_rle/16/8         1.67 ns         1.67 ns    419700051
compare256_rle/16/64        11.4 ns         11.4 ns     61416695
compare256_rle/16/256       44.8 ns         44.8 ns     15637194
compare256_rle/32/1         1.39 ns         1.39 ns    503483411
compare256_rle/32/8         1.95 ns         1.95 ns    359652651
compare256_rle/32/64        6.43 ns         6.41 ns    109126696
compare256_rle/32/256       20.1 ns         20.0 ns     34895172
compare256_rle/64/1        0.696 ns        0.695 ns   1000000000
compare256_rle/64/8        0.697 ns        0.696 ns   1000000000
compare256_rle/64/64        2.79 ns         2.79 ns    250897664
compare256_rle/64/256       9.59 ns         9.58 ns     73801000
```

ccawley2011 commented:
> ### Compare256 comparison benchmarks
> #### Aarch64

Are you able to get benchmarks on the same machine when building with `-mstrict-align`?

Dead2 commented:
> Are you able to get benchmarks on the same machine when building with `-mstrict-align`?

Sure.
```
compare256/8/1              1.04 ns         1.04 ns    612475649
compare256/8/8              4.70 ns         4.69 ns    149093282
compare256/8/64             35.4 ns         35.4 ns     19790512
compare256/8/256             141 ns          141 ns      4949127
compare256/16/1             1.46 ns         1.46 ns    479843890
compare256/16/8             2.94 ns         2.94 ns    238345836
compare256/16/64            19.0 ns         19.0 ns     36641896
compare256/16/256           74.0 ns         74.0 ns      9461339
compare256/32/1             1.25 ns         1.25 ns    559811241
compare256/32/8             2.08 ns         2.08 ns    335893024
compare256/32/64            15.4 ns         15.4 ns     45391289
compare256/32/256           55.4 ns         55.4 ns     12627351
compare256/64/1             1.39 ns         1.39 ns    503843948
compare256/64/8             1.39 ns         1.39 ns    503831186
compare256/64/64            8.75 ns         8.75 ns     79974087
compare256/64/256           28.8 ns         28.8 ns     24339608
compare256/neon/1           10.1 ns         10.1 ns     69423400
compare256/neon/8           10.1 ns         10.1 ns     69453530
compare256/neon/64          37.4 ns         37.4 ns     18712361
compare256/neon/256          146 ns          146 ns      4808873
compare256_rle/8/1          1.25 ns         1.25 ns    559811988
compare256_rle/8/8          3.75 ns         3.75 ns    186607782
compare256_rle/8/64         30.0 ns         30.0 ns     23325738
compare256_rle/8/256         120 ns          120 ns      5831450
compare256_rle/16/1         1.25 ns         1.25 ns    559815136
compare256_rle/16/8         2.29 ns         2.29 ns    298684297
compare256_rle/16/64        16.7 ns         16.7 ns     41986882
compare256_rle/16/256       66.7 ns         66.7 ns     10496519
compare256_rle/32/1         1.25 ns         1.25 ns    559824672
compare256_rle/32/8         3.75 ns         3.75 ns    186607377
compare256_rle/32/64        15.4 ns         15.4 ns     45390761
compare256_rle/32/256       55.4 ns         55.4 ns     12627266
compare256_rle/64/1         1.56 ns         1.56 ns    479850288
compare256_rle/64/8         1.47 ns         1.47 ns    444989578
compare256_rle/64/64        7.09 ns         7.09 ns     98790625
compare256_rle/64/256       27.1 ns         27.1 ns     25838011
```
The 16bit variant is still clearly faster on Aarch64 with `-mstrict-align` set, in fact I don't see any significant changes at all.

ccawley2011 commented:
I ran the benchmark on a Raspberry Pi 1B with and without `-mno-unaligned-access`, and am getting different results for each one:

Default:
```
compare256/c/1                                38.1 ns         38.1 ns     18385008
compare256/c/8                                96.8 ns         96.6 ns      7228866
compare256/c/64                                682 ns          611 ns      1147296
compare256/c/256                              2346 ns         2335 ns       298880
compare256/unaligned_16/1                     39.7 ns         39.6 ns     17708787
compare256/unaligned_16/8                     64.6 ns         64.4 ns     10847163
compare256/unaligned_16/64                     347 ns          342 ns      2052322
compare256/unaligned_16/256                   1263 ns         1257 ns       555961
compare256/unaligned_32/1                     38.3 ns         38.1 ns     18378588
compare256/unaligned_32/8                     60.1 ns         60.0 ns     11649774
compare256/unaligned_32/64                     226 ns          226 ns      3103525
compare256/unaligned_32/256                    790 ns          788 ns       888088
```

-mno-unaligned-access:
```
compare256/c/1                                38.1 ns         38.1 ns     18391336
compare256/c/8                                96.8 ns         96.6 ns      7228418
compare256/c/64                                610 ns          609 ns      1148313
compare256/c/256                              2334 ns         2331 ns       300314
compare256/unaligned_16/1                      252 ns          252 ns      2777095
compare256/unaligned_16/8                      920 ns          916 ns       764234
compare256/unaligned_16/64                    7121 ns         7109 ns        98341
compare256/unaligned_16/256                  28357 ns        28318 ns        24722
compare256/unaligned_32/1                      201 ns          201 ns      3485309
compare256/unaligned_32/8                      379 ns          378 ns      1851627
compare256/unaligned_32/64                    2720 ns         2717 ns       257620
compare256/unaligned_32/256                  10752 ns        10735 ns        65128
```

KungFuJesus commented:
Yeah @ccawley2011's results on arm with forced alignment requirements are enough for me to say for architectures that have no unaligned access we need the 8 bit variant. Wow, those memcpys add up.

But also luckily, that is mostly a compile time decision.

Dead2 commented:
@ccawley2011 Could you run `build/test/benchmarks/benchmark_zlib --benchmark_filter=compress*` on that Rpi1 before and after this PR, both with `-mno-unaligned-access`.
I am curious about the effects from changes in `match_tpl.h`, hopefully those are not as bad as the ones you posted earlier.
I also wonder how/whether your PR #1548 affects that on top of this PR.

Dead2 commented:
Rebased, fixed missing `LONGEST_MATCH_SLOW` define, merged checks in `compare256_c.c`, and cleaned up some minor discrepancies between `compare256_p.h` and `compare256_rle.h`.

KungFuJesus commented:
WITH_NATIVE fails on PPC64:
```
/home/adam/zlib-ng/deflate_fast.c:55:29: error: call to undeclared function 'longest_match_generic'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   55 |                 match_len = FUNCTABLE_CALL(longest_match)(s, hash_head);
      |                             ^
/home/adam/zlib-ng/functable.h:21:32: note: expanded from macro 'FUNCTABLE_CALL'
   21 | #  define FUNCTABLE_CALL(name) native_ ## name
      |                                ^
<scratch space>:229:1: note: expanded from here
  229 | native_longest_match
      | ^
/home/adam/zlib-ng/arch/generic/generic_functions.h:52:32: note: expanded from macro 'native_longest_match'
   52 | #  define native_longest_match longest_match_generic
      |                                ^
1 error generated.
make[2]: *** [CMakeFiles/zlib.dir/build.make:244: CMakeFiles/zlib.dir/deflate_fast.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

Dead2 commented:
> WITH_NATIVE fails on PPC64

I think this is fixed now.

phprus commented:
Workflow ``CMake / EL9 Clang S390X DFLTCC MSAN`` is failing.

https://github.com/zlib-ng/zlib-ng/actions/runs/12470358964/job/34805368017?pr=1830

Dead2 commented:
> Workflow `CMake / EL9 Clang S390X DFLTCC MSAN` is failing.

Yes, it also fails before this PR, we haven't had that test working for months due to github not wanting to support s390x themselves, so we have to depend on a 3rd party runner port. Not sure what broke there yet.
I have pinged @iii-i, but he is likely off on a Christmas vacation or something I would guess.

KungFuJesus commented:
> > WITH_NATIVE fails on PPC64
> 
> I think this is fixed now.

Confirmed.  Works now.

Dead2 commented:
Rebased and added a couple comments improving documentation.

Dead2 commented:
After talking it over with @KungFuJesus, we agree that checking for OPTIMAL_CMP is probably not necessary or useful in zmemory.h. If it turns out that it is, it can be added back in a fixed form. I was considering changing them to all be => 16, but we'll try it like this first and see what tests and benchmarks come out as.

@ccawley2011 Please have a look at the updated PR now that your PR #1548 has been merged first. It seems to work fine with `-mstrict-align` on my 64-bit rpi5 at least, but that supports `HAVE_MAY_ALIAS` so it isn't actully affected by the change.

Dead2 commented:
Fixed indentation.

ccawley2011 commented:
Is there an easy way to get benchmarks for the LONGEST_MATCH functions?

Dead2 commented:
> Is there an easy way to get benchmarks for the LONGEST_MATCH functions?

I think the closest you can get currently is the gbench compress* benchmarks.
`test/benchmarks/benchmark_zlib --benchmark_filter=compress*`
